### ISSUE TITLE
investigate what differentiates high-accuracy rollouts from average-accuracy rollouts across targets

### ISSUE THREAD

**Issue by timodonnell** [Open-Athena/MarinFold#102]
We typically see a pretty big range of contact prediction accuracies across rollouts for a target protein. For example on the right of the plot below is the distribution of F1 scores for n=1000 rollouts for a single target protein from our contacts-v1 training set (this is from experiment #98):

<img width="1009" height="368" alt="Image" src="https://github.com/user-attachments/assets/52853479-2db3-455a-a756-135501056b83" />

It would be helpful if we can glean anything about what differentiates high accuracy rollouts from average accuracy rollouts across targets. For example stuff like: are the high accuracy rollouts just longer? is there some bias in the order that contacts are predicted for high accuracy rollouts (e.g. long range first, short range first, high perplexity first, certain kinds of amino acids first, something else)?

Relatedly but maybe harder to answer - within a single protein target, what drives the variation across rollouts? For example, is there a single highly informative contact, and the high accuracy rollouts are the ones that guess that early on?

For this we should work with the current best contacts-v1 model, i.e. the one at https://huggingface.co/buckets/open-athena/MarinFold/tree/checkpoints/prot-exp75-cv1-1_5b-e8-lr1e-3-wd0p2-v1-bc3084/hf/step-35679

**Comment by allensu02** [comment]
Thanks Tim for pointing me towards this problem, this looks interesting and I'll take a look.

**Comment by timodonnell** [comment]
Sounds great @allensu02 . I generated some data you can use to get started here - take a look at [exp102](https://github.com/Open-Athena/MarinFold/tree/main/experiments/exp102_rollout_accuracy_factors).

**Comment by allensu02** [comment]
**Progress update for #102**

I analyzed the exp102 ordered-rollout data to understand what separates
high-F1 rollouts from average rollouts for the same protein.

I started from the starter analysis in exp102. That analysis found a weak
within-target relationship between rollout F1 and output length/contact count,
little generic short/medium/long emission-order bias, and a striking example
where one contact's "earliness" correlated strongly with F1. I treated those as
hypotheses rather than final results: length needed to be separated from contact
quality, and the early-contact example needed to separate "the contact appeared
at all" from "the contact appeared early." That led to the checks below.

**Completed checks**

- **Key contacts: presence, order, and confidence.** I tested whether one
  important true contact for a target can identify better rollouts. To avoid a
  circular result, I picked the contact using 500 discovery rollouts, then
  tested it on the other 500 rollouts using F1 with that contact removed from
  the score. Rollouts that contained the selected contact had better remaining
  contact maps by **+0.0560 F1** (95% target-bootstrap CI **[0.0491, 0.0636]**),
  with positive effects for 99.5% of eligible targets. The contact's local
  logprob also had signal. Raw earliness was weak (rho **+0.0259**), so I
  would not yet claim that emitting a contact early is itself the important
  factor.

- **Length / number of emitted contacts.** High-F1 rollouts emitted slightly
  more valid contacts (**+2.342** on average), but this did not explain the
  accuracy gap. After matching rollouts to have almost the same number of
  predicted contacts, high rollouts still had about **+0.0796 F1** (CI
  **[0.0728, 0.0868]**). Thus, contact quantity alone does not explain the
  difference; high-F1 rollouts appear to choose better contacts.

- **Agreement with independent consensus.** I built a consensus contact set
  from one half of each target's rollouts and scored the other half against it,
  so no rollout was compared to a consensus that included itself. High-F1
  rollouts had higher independent top-K Jaccard by **+0.0355** (CI
  **[0.0307, 0.0406]**), and the effect stayed about **+0.0341** after matching
  for contact count. High rollouts also contained about **+2.4** useful
  low-frequency true contacts, so pure consensus does not capture every useful
  edge.

- **Consensus as an actual rollout selector.** I then tested whether the
  consensus signal can rank individual rollouts, not just separate high-F1 and
  average-F1 groups after the fact. Top-K Jaccard with independent consensus
  reached ROC-AUC **0.6913** (CI **[0.6750, 0.7080]**) and Spearman **0.4902**
  (CI **[0.4559, 0.5234]**). Selecting the top rollout by this score improved
  F1 by **+0.0760** over random (CI **[+0.0648, +0.0875]**). This makes
  consensus the strongest label-free selector baseline so far, though the
  selected rollout remains **0.1346 F1** below the oracle on average and the
  score requires a separate target-specific reference ensemble.

- **Consensus sample efficiency.** I repeated the no-self consensus analysis
  from 1 to 500 references using frozen nested subsets and a deterministic
  rerun. **100 reference rollouts retained nearly all of the 500-reference
  consensus signal under the preregistered saturation test**: Jaccard effect
  **+0.0340** versus **+0.0355**, AUC **0.6880** versus **0.6913**, and top-1
  lift **+0.0764** versus about **+0.0760**. N=50 was close but failed the
  frozen 90%-equivalence and next-step-gain gates, so it is not called
  saturated.

- **Topology / map-shape features.** I tested whether better rollouts have
  more coherent contact-map structure after controlling for contact count. The
  broad version of this idea did not hold up: better maps were not simply more
  globally connected or coherent. A few narrow local features did move in the
  expected direction, such as adjusted parallel continuity (**+0.0064**, CI
  **[+0.0041, +0.0087]**) and lower strict endpoint-interleaving density
  (**-0.0058**, CI **[-0.0080, -0.0035]**), but these are small auxiliary
  candidates rather than a standalone explanation. Largest-component fraction
  reversed from raw +0.0041 to adjusted/exact about -0.0025, while adjusted
  component count increased by +0.0915, directly contradicting blanket
  "more connected" language.

**Current takeaway**

High rollouts seem better because they select a better contact set. They agree
more with independently sampled model support, but they also sometimes include
rare correct contacts that simple voting may undervalue. Local contact
confidence and a few map-shape features may help, but whole-rollout NLL was
weak in exp98 and raw order is not yet convincing. For future consensus work
whose design is not already frozen, 100 references is now the default.

**Remaining questions**

- Test whether combining consensus agreement with local map-shape features
  improves rollout selection.
- Check whether consensus sometimes misses rare but correct contacts because it
  follows the majority.
- Compare how many reference rollouts are needed to build a useful consensus at
  a fixed generation cost. The current analysis suggests that roughly 100
  references may capture most of the signal seen with 500, but this needs
  further testing.

**Caveats**

The key-contact test used ground truth to choose the contact, so it is a
characterization result rather than a deployable selector. The consensus
selector is label-free, but it still needs a target-specific ensemble of
rollouts and has only been tested retrospectively on these contacts-v1 training
targets. The topology effects are small, no practical magnitude threshold was
preregistered, and strict endpoint interleaving is combinatorial rather than a
physical-invalidity test. Current count controls are observational rather than
causal. E3S's N=100 result is relative to this retrospective exp102 Jaccard
rule, not a universal optimum; ten deterministic subsets do not exhaust all
reference samples, and selection remains about 0.134 F1 below oracle. External
claims should eventually be checked on structurally disjoint held-out proteins.
